keps/sig-cluster-lifecycle/0015-kubeadm-join-master.md

```diff
@@ -0,0 +1,431 @@
+            - [The kubeadm join --master workflow](#the-kubeadm-join---master-workflow)
+            - [dynamic workflow (advertise-address == `controlplaneAddress`)](#dynamic-workflow-advertise-address--controlplaneaddress)
+            - [Static workflow (advertise-address != `controlplaneAddress`)](#static-workflow-advertise-address--controlplaneaddress)
+            - [Strategies for deploying control plane components](#strategies-for-deploying-control-plane-components)
+            - [Strategies for distributing cluster certificates](#strategies-for-distributing-cluster-certificates)
+            - [`kubeadm upgrade` for HA clusters](#kubeadm-upgrade-for-ha-clusters)
+    - [Graduation Criteria](#graduation-criteria)
+    - [Implementation History](#implementation-history)
+    - [Drawbacks](#drawbacks)
+    - [Alternatives](#alternatives)
+
+<!-- /TOC -->
+
+## Summary
+
+We are extending the kubeadm distinctive `init` and `join` workflow, introducing the
+capability to add more than one master node to an existing cluster by means of the
+new `kubeadm join --master` option (in alpha release the flag will be named --experimental-master)
+
+As a consequence, kubeadm will provide a best-practice, “fast path” for creating a
+minimum viable, conformant Kubernetes cluster with one or more master nodes and
+zero or more worker nodes.
+
+## Motivation
+
+Support for high availability is one of the most requested features for kubeadm.
+
+Even if, as of today, there is already the possibility to create an HA cluster
+using kubeadm in combination with some scripts and/or automation tools (e.g.
+[this](https://kubernetes.io/docs/setup/independent/high-availability/)), this KEP was
+designed with the objective to introduce an upstream simple and reliable solution for
+achieving the same goal.
+
+Such solution will provide a consistent and repeatable base for implementing additional
+capabilities like e.g. kubeadm upgrade for HA clusters.
+
+### Goals
+
+- "Divide and conquer”
+
+  This proposal - at least in its initial release - does not address all the possible
+  user stories for creating an highly available Kubernetes cluster, but instead
+  focuses on:
+
+  - Defining a generic and extensible flow for bootstrapping an HA cluster, the
+    `kubeadm join --master` workflow.
+  - Providing a solution *only* for well defined user stories. see
+    [User Stories](#user-stories) and [Non-goals](#non-goals).
+
+- Enable higher-level tools integration
+
+  We expect higher-level and tooling will leverage on kubeadm for creating HA clusters;
+  accordingly, the `kubeadm join --master` workflow should provide support for
+  the following operational practices used by higher level tools:
+
+  - Parallel node creation
+
+    Higher-level tools could create nodes in parallel (both masters and workers)
+    for reducing the overall cluster startup time.
+    `kubeadm join --master` should support natively this practice without requiring
```

Comment:
and how does this work?

Reply:
You can create in parallel bootstrap master, secondary masters and workers because:
1. all the joining nodes (secondary masters and workers) use the same discovery mechanism, that "automatically" waits for the bootstrap master to complete its setup
2. the joining nodes (secondary masters and workers) can join in any order